cli/command/image/build.go

```diff
@@ -268,7 +308,59 @@ func runBuild(dockerCli *command.DockerCli, options buildOptions) error {
 		progressOutput = &lastProgressOutput{output: progressOutput}
 	}
 
-	var body io.Reader = progress.NewProgressReader(buildCtx, progressOutput, 0, "", "Sending build context to Docker daemon")
+	if dockerfileCtx != nil && buildCtx == nil {
+		buildCtx = dockerfileCtx
+	}
+	var s *session.Session
+	if isSessionSupported(dockerCli) {
```

Comment:
This function is already way too large. There's no way we can probably unit test most of it. We really should not be adding any significant code to it.

Could we make the rest of this (up to line 362, and the `if s != nil ` block) into a few new functions? Even if the arg list is long, at this point I think it's still an improvement. We can work on refactoring args into structs in a follow up.

Maybe even a new file along with `sizeProgress` and `bufferedWriter` ? That way all the session handling stuff is together and not mixed up with other build things.